### ISSUE TITLE
Ph.D. '17 | Variational Inference and Deep Learning: A New Synthesis.

### ISSUE THREAD

**Issue by NorbertZheng** [NorbertZheng/read-papers#21]
Kingma, Diederik Pieter. [Variational inference & deep learning: A new synthesis](https://pure.uva.nl/ws/files/17891313/Thesis.pdf).

**Comment by NorbertZheng** [comment]
## Related Reference
 - Joseph Rocca. [Understanding Variational Autoencoders (VAEs)](https://towardsdatascience.com/understanding-variational-autoencoders-vaes-f70510919f73).

**Comment by NorbertZheng** [comment]
## Introduction

**Comment by NorbertZheng** [comment]
## Dimensionality reduction, PCA and autoencoders

**Comment by NorbertZheng** [comment]
## Variational Autoencoders

**Comment by NorbertZheng** [comment]
### Intuitions about the regularisation
**The regularity that is expected from the latent space** in order to make generative process possible can be expressed through two main properties:
 - **Continuity**: Two close points in the latent space should not give two completely different contents once decoded.
 - **Completeness**: For a chosen distribution, a point sampled from the latent space should give "meaningful" content once decoded.

![image](https://user-images.githubusercontent.com/37235023/156872567-e6e4638b-3151-44e7-9cb1-6839c58dd9e1.png)

The only fact that **VAEs encode inputs as distributions instead of simple points** is not sufficient to ensure continuity and completeness. Without a well defined regularization term, the model can learn, in order to minimize its reconstruction error, **to "ignore" the fact that distributions are returned and behave almost like classic auto-encoders** (leading to over-fitting). To do so, the encoder can
 - Return distributions with tiny variances (that would tend to be punctual distributions).
 - Return distributions with very different means (that would then be really far apart from each other in the latent space).

In both cases, distributions are used the wrong way (cancelling the expected benefit) and continuity and/or completeness are not satisfied.

So, in order to avoid these effects we have to **regularize both the covariance matrix and the mean of the distributions returned by the encoder**. In practice, this regularization is done by enforcing distributions to be close to a **standard normal distribution** (centered and reduced). This way, we require
 - The covariance matrices to be close to the identity, preventing punctual distributions.
 - The mean to be close to 0, preventing encoded distributions to be too far apart from each others.

![image](https://user-images.githubusercontent.com/37235023/156872888-3a75963a-e4f0-4b86-af2c-550e9162abfd.png)

With this regularization term, we prevent the model to encode data far apart in the latent space and encourage as much as possible returned distributions to "overlap", satisfying this way the expected continuity and completeness conditions. **Naturally, as for any regularization term, this comes at the price of a higher reconstruction error on the training data**. The trade-off between the reconstruction error and the KL-divergence can however be adjusted and we will see in the next section how the expression of the balance naturally emerge from our formal derivation.

To conclude, we can observe that continuity and completeness obtained with regularization **tend to create a "gradient" over the information encoded in the latent space**. For example, a point of the latent space that would be halfway between the means of two encoded distributions coming from different training data should be decoded in something that is somewhere between the data that gave the first distribution and the data that gave the second distribution as it may be sampled by the auto-encoder in both cases.
![image](https://user-images.githubusercontent.com/37235023/156873216-147fa4cf-58fb-4c27-bc21-c0a4de2442b8.png)

This reminds me of a continuous attractor neural network (CANN), there is a continuous set of attractors in the latent space, which corresponds to the gradient over the information encoded in the latent space, satisfying the **continuity** property of VAEs.

**Note**. As a side note, we can mention that the second potential problem we have mentioned (the network put distributions far from each others) is in fact almost equivalent to the first one (the network tends to return punctual distribution) up to a change of scale: in both case variances of distributions become small relatively to distance between their means.

**Comment by NorbertZheng** [comment]
### Intuitions about the regularisation
**The regularity that is expected from the latent space** in order to make generative process possible can be expressed through two main properties:
 - **Continuity**: Two close points in the latent space should not give two completely different contents once decoded.
 - **Completeness**: For a chosen distribution, a point sampled from the latent space should give "meaningful" content once decoded.

![image](https://user-images.githubusercontent.com/37235023/156872567-e6e4638b-3151-44e7-9cb1-6839c58dd9e1.png)

The only fact that **VAEs encode inputs as distributions instead of simple points** is not sufficient to ensure continuity and completeness. Without a well defined regularization term, the model can learn, in order to minimize its reconstruction error, **to "ignore" the fact that distributions are returned and behave almost like classic auto-encoders** (leading to over-fitting). To do so, the encoder can
 - Return distributions with tiny variances (that would tend to be punctual distributions).
 - Return distributions with very different means (that would then be really far apart from each other in the latent space).

In both cases, distributions are used the wrong way (cancelling the expected benefit) and continuity and/or completeness are not satisfied.

So, in order to avoid these effects we have to **regularize both the covariance matrix and the mean of the distributions returned by the encoder**. In practice, this regularization is done by enforcing distributions to be close to a **standard normal distribution** (centered and reduced). This way, we require
 - The covariance matrices to be close to the identity, preventing punctual distributions.
 - The mean to be close to 0, preventing encoded distributions to be too far apart from each others.

![image](https://user-images.githubusercontent.com/37235023/156872888-3a75963a-e4f0-4b86-af2c-550e9162abfd.png)

With this regularization term, we prevent the model to encode data far apart in the latent space and encourage as much as possible returned distributions to "overlap", satisfying this way the expected continuity and completeness conditions. **Naturally, as for any regularization term, this comes at the price of a higher reconstruction error on the training data**. The trade-off between the reconstruction error and the KL-divergence can however be adjusted and we will see in the next section how the expression of the balance naturally emerge from our formal derivation.

To conclude, we can observe that continuity and completeness obtained with regularization **tend to create a "gradient" over the information encoded in the latent space**. For example, a point of the latent space that would be halfway between the means of two encoded distributions coming from different training data should be decoded in something that is somewhere between the data that gave the first distribution and the data that gave the second distribution as it may be sampled by the auto-encoder in both cases.
![image](https://user-images.githubusercontent.com/37235023/156873216-147fa4cf-58fb-4c27-bc21-c0a4de2442b8.png)

This reminds me of a continuous attractor neural network (CANN), there is a continuous set of attractors in the latent space, which corresponds to the gradient over the information encoded in the latent space, satisfying the **continuity** property of VAEs.

**Note**. As a side note, we can mention that the second potential problem we have mentioned (the network put distributions far from each others) is in fact almost equivalent to the first one (the network tends to return punctual distribution) up to a change of scale: in both case variances of distributions become small relatively to distance between their means.

**Comment by NorbertZheng** [comment]
## Mathematical details of VAEs
In the previous section we gave the following intuitive overview:
 - VAEs are autoencoders that encode inputs as distributions instead of points and whose latent space "organization" is regularized by constraining distributions returned by the encoder to be close to a standard Gaussian. In this section we will give a more mathematical view of VAEs that will allow us to **justify the regularization term more rigorously. To do so, we will set a clear probabilistic framework and will use, in particular, variational inference technique.
### Probabilistic framework and assumptions
Let's begin by defining a probabilistic graphical model to describe our data. We denote by x the variable that represents our data and assume that x is generated from a latent variable z (the encoded representation) that is not directly observed. Thus, for each data point, the following two steps generative process is assumed:
 - First, a latent representation z is sampled from the prior distribution p(z).
 - Second, the data x is sampled from the conditional likelihood distribution p(x|z).

![image](https://user-images.githubusercontent.com/37235023/156926028-ac18c983-6687-4b7a-9495-39b3a2900c0c.png)
With such a probabilistic model in mind, we can redefine our notions of encoder and decoder. Indeed, contrarily to a simple autoencoder that consider determinisitic encoder and decoder, we are going to consider now **probabilistic versions of these two objects**.
 - **Probabilistic decoder**: p(x|z), that describes the distribution of the decoded variable given the encoded one.
 - **Probabilistic encoder**: p(z|x), that describes the distribution of the encoded variable given the decoded one.

At this point, we can already notice that the regularization of the latent space that we lacked in simple autoencoders naturally appears here in the definition of the data generation process: **encoded representations z in the latent space are indeed assumed to follow the prior distribution p(z)**. Otherwise, we can also remind the well-known Bayes theorem that makes the link between the prior p(z), the likelihood p(x|z), and the posterior p(z|x):
![image](https://user-images.githubusercontent.com/37235023/156926813-afcb775c-bc1a-4006-9ace-ad7353243bab.png)

Let's now make the assumption (regularization):
 - p(z) is a standard Gaussian distribution.
 - p(x|z) is a Gaussian distribution:
    - The mean is defined by a **deterministic** function f of the variable z.
    - The covariance matrix has the form of a positive constant c that multiplies the identity matrix I.

The function f is assumed to belong to a family of functions denoted F that is left unspecified for the moment and that will be chosen later. Thus, we have
![image](https://user-images.githubusercontent.com/37235023/156926818-eb48a114-1e41-4aed-b450-2faec228df3c.png)

Let's consider, for now, that f is well defined and fixed. In theory, as we know p(z) and p(x|z), we can use the Bayes theorem to compute p(z|x): this is a classical [Bayesian inference problem](https://towardsdatascience.com/bayesian-inference-problem-mcmc-and-variational-inference-25a8aa9bce29). However, as we discussed in our previous article, this kind of computation is often **intractable (because of the integral at the denominator)** and require the use of approximation techniques such as variational inference.

**Note**. Here we can mention that p(z) and p(x|z) are both Gaussian distribution. So, if we had E(x|z)=f(z)=z, it would imply that p(z|x) should also follow a Gaussian distribution and, in theory, we could "only" try to express the mean and the covariance matrix of p(z|x) with respect to the means and the covariance matrices of p(z) and p(x|z). However, in practice this condition (e.g. E(x|z)=f(z)=z) is not met and we need to use of an approximation technique like variational inference that makes the approach pretty general and more robust to some changes in the hypothesis of the model (generative model).

**Comment by NorbertZheng** [comment]
### Variational inference formulation
In statistics, **variational inference(VI) is a technique to approximate complex distributions**. The idea is to set a **parameterized** family of distribution (for example the family of Gaussians, whose parameters are the mean and the covariance) and to look for the best approximation of our target distribution among this family. The best element in the family is one that **minimize a given approximation error measurement** (most of the time the Kullback-Leibler divergence between approximation and target) and is found by gradient descent over the parameters that describe the family. Please refer to [Variational Inference](https://towardsdatascience.com/bayesian-inference-problem-mcmc-and-variational-inference-25a8aa9bce29) for more details.
Here we are going to approximate `p(z|x)` by a Gaussian distribution `q_x(z)` whose mean and covariance are defined by two functions, `g` and `h`, of the parameter `x`. These two functions are supposed to belong, respectively, to the families of the functions `G` and `H` that will be specified later but that are supposed to be parameterized. Thus we can denote
![image](https://user-images.githubusercontent.com/37235023/156956066-2a099c10-2913-47b5-8638-34dd5a7327ca.png)

So, we have defined this way a family of candidates for variational inference and need now to find the best approximation among this family by optimizing the functions `g` and `h` (in fact, their parameters) to minimize the Kullback-Leibler divergence between the approximation and the target `p(z|x)`. In other words, we are looking for the optimal `g*` and `h*` such that
![image](https://user-images.githubusercontent.com/37235023/156957010-bd790d67-50a9-4436-91fb-7283abd2528b.png)

In the second last equation, we can observe the trade-off there exists, when approximating the posterior `p(z|x)`, between 
 - Maximizing the likelihood of the "observation" (maximization of the expected log-likelihood, for the first term).
 - Staying close to the prior distribution (minimization of the KL-divergence between `q_x(z)` and `p(z)`, for the second term).

**This trade-off is natural for Bayesian inference problem and express the balance that needs to be found between the confidence we have in the data and the confidence we have in the prior.**

Up to know, we have assumed the function `f` known and fixed and we have showed that, under such assumptions, we can approximate the posterior `p(z|x)` using variational inference technique. However, in practice this function `f`, that defines the decoder, is not known and also need to be chosen. To do so, let's remind that our initial goal is to find a performant encoding-decoding scheme whose latent space is regular enough to be used for generative purpose. If the regularity is mostly ruled by the prior distribution assumed over the latent space, the performance of the overall encoding-decoding scheme highly depends on the choice of the function `f`. Indeed, as `p(z|x)` can be approximate (by variational inference) from `p(z)` and `p(x|z)` and as `p(x)` is a simple standard Gaussian, the only two levers we have at our disposal in our model to make optimizations are
 - **The parameter `c`**: Defines the variance of the likelihood.
 - **The function `f`**: Defines the mean of the likelihood.

**Comment by NorbertZheng** [comment]
### Bringing neural networks into the model
Up to now, we have set a probabilistic model that depends on three functions, `f`, `g` and `h`, and express, using variational inference, the optimization problem to solve in order to get `f*`, `g*` and `h*` that give the optimal encoding-decoding scheme with this model. **As we can't easily optimize over the entire space of functions, we constrain the optimization domain  and decide to express `f`, `g` and `h` as neural networks.** Thus, `F`, `G` and `H` correspond respectively to the families of functions defined by the networks architectures and optimization is done over the parameters of these networks.

In practice, `g` and `h` are not defined by two completely independent networks but share a part of their architecture and their weights so that we have
![image](https://user-images.githubusercontent.com/37235023/156962287-5f814c4a-5738-4e9e-a252-2cb097b3389f.png)

As it defines the covariance matrix of `q_x(z)`, `h(x)` is supposed to be a square matrix. However, in order to simplify the computation and reduce the number of parameters, we make the additional assumption that our approximation of `p(z|x)`, `q_x(z)`, is a multidimensional Gaussian distribution with diagonal covariance matrix (variables independence assumption). With this assumption, `h(x)` is simply the vector of the diagonal elements of the covariance matrix and has then the same size as `g(x)`. However, we reduce this way the family of distributions we consider for variational inference and, so, the approximation of `p(z|x)` obtained can be less accurate.
![image](https://user-images.githubusercontent.com/37235023/156962699-5082b74d-cf9f-48a2-98e3-db358eea47e0.png)

Contrarily to the encoder part that models `p(z|x)` and for which we considered a Gaussian with both mean and covariance that are functions of `x` (`g` and `h`), our model assumes for `p(x|z)` a Gaussian with fixed covariance. The function `f` of the variable `z` defining the mean of that Gaussian is modelled by a neural network and can be represented as follows:
![image](https://user-images.githubusercontent.com/37235023/156962966-bfa59332-7cc0-4102-b20d-8230391c24a3.png)

The overall architecture is then obtained by concatenating the encoder and the decoder parts. However we still need to be very careful about the way we sample from the distribution returned by the encoder during the training. The sampling process has to be expressed in a way that allows the error to be backpropagated through the network. A simple trick, called **representation trick**, is used to make the gradient descent possible despite the random sampling that occurs halfway of the architecture and consists in using the fact that if `z` is a random variable following a Gaussian distribution with mean `g(x)` and with covariance `H(x)=h(x).h^t(x)` then it can be expressed as
![image](https://user-images.githubusercontent.com/37235023/156963346-b94d103c-92a1-4ee5-ba51-5fe411e6c6b1.png)
![image](https://user-images.githubusercontent.com/37235023/156963470-e2d7b864-7c5b-4069-9e6f-28e6bf5fa32e.png)

Finally, the objective function of the variational autoencoder architecture obtained this way is given by the last equation of the previous subsection in which the theoretical expectancy is replaced by a more or less accurate Monte-Carlo approximation that consists, most of the time, into a single draw. So, considering this approximation and denoting `C=1/(2c)`, we recover the loss function derived intuitively in the previous section, composed of a reconstruction term, a regularizaion term and a constant to define the relative weights of these two terms.
![image](https://user-images.githubusercontent.com/37235023/156963773-3fde6e53-df6b-4478-9f3b-a2bf7656aa60.png)

**Comment by NorbertZheng** [comment]
## Takeaways
The main takeaways of thie article are:
 - Dimensionality reduction is the process of reducing the number of features that describe some data (either by selecting only a subset of the initial features or by combining them into a reduced number new features) and, so, can be seen as an encoding process.
 - Autoencoders are neural networks architectures composed of both an encoder and a decoder that create a bottleneck to go through for data and that are trained to lose a minimal quantity of information during the encoding-decoding process (training bt gradient descent iterations with the goal to reduce the reconstruction error).
 - Due to over-fitting, the latent space of an autoencoder can be extremely irregular (close points in latent space can give very different decoded data, some point of the latent space can give meaningless content once decoded, ...) and, so, we can't really define a generative process that simply consists to sample a point from the latent space and make it go through the decoder to get a new data.
 - Variational AutoEncoders(VAEs) are autoencoders that tackle the problem of the latent space irregularity by making the encoder return a distribution over the latent space instead of a single point and by adding in the loss function a regularization term over the returned distribution in order to ensure a better organization of the latent space.
 - Assuming a simple underlying probabilistic model to describe our data, the pretty intuitive loss function of VAEs, composed of a reconstruction term and a regularization term, can be carefully derived, using in particular the statistical technique of variational inference (hence the name "variational" autoencoders).

To conclude, we can outline that, during the last years, GANs have benefited from much more scientific contributions than VAEs. Among other reasons, the higher interest that has been shown by the community for GANs can be partly explained by the higher degree of complexity in VAEs theoretical basis (probabilistic model and variational inference) compared to the simplicity of the adversarial training concept that rules GANs.
